bpo-1635741: Convert _sre types to heap types and establish module state (PEP 384)

erlend-aasland opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-1635741](https://bugs.python.org/issue1635741) -->
https://bugs.python.org/issue1635741
<!-- /issue-number -->

Commit:
Convert _sre pattern type to heap type

Commit:
Add convenience macro

Commit:
Convert _sre match type to heap type

Commit:
Convert _sre scanner type to heap type

Commit:
Establish global module state and add types to it

Commit:
Convert global state to module state

Commit:
Convert _sre to multi-phase initialisation

Commit:
Add NEWS

tiran commented on Modules/_sre.c:
```diff
@@ -248,22 +248,38 @@ data_stack_grow(SRE_STATE* state, Py_ssize_t size)
 
+static struct PyModuleDef sremodule;
+#define get_sre_module_state_by_type(tp) \
+    (get_sre_module_state(_PyType_GetModuleByDef(tp, &sremodule)))
```
None of the types in the module have ``Py_TPFLAGS_BASETYPE``. AFAIK only subclass-able types need ``_PyType_GetModuleByDef``.

erlend-aasland replied:
This is used to fetch the state from slot methods. Is there a better way?

tiran replied:
I'm convinced that it's safe to use ``get_sre_module_state_by_class`` everywhere in this module. The left hand argument of ``richcompare`` is always an instance of the type. Python does not let users pass another type. The argument clinic lines are safe to use ``PyType_GetModule()``, too.

``_PyType_GetModuleByDef()`` is only required when a C extension type has the ``Py_TPFLAGS_BASETYPE`` flag. Without the basetype flag users cannot subclass an extension class and ``PyType_GetModule()`` will always return the module from ``PyType_FromModuleAndSpec``.

erlend-aasland replied:
> `_PyType_GetModuleByDef()` is only required when a C extension type has the `Py_TPFLAGS_BASETYPE` flag. Without the basetype flag users cannot subclass an extension class and `PyType_GetModule()` will always return the module from `PyType_FromModuleAndSpec`.

Got it; thanks. Fixed in bdd133884d

tiran commented on Modules/_sre.c:
```diff
@@ -248,22 +248,38 @@ data_stack_grow(SRE_STATE* state, Py_ssize_t size)
+    PyTypeObject *Pattern_Type;
+    PyTypeObject *Match_Type;
+    PyTypeObject *Scanner_Type;
+} sre_module_state;
```
I would prefer if you stick to common naming conventions:

* name the state after the module: ``_sremodulestate``
* name the state variable in functions ``state``, not ``module_state``

erlend-aasland replied:
The problem is that `_sre` already has a "SRE state", which has nothing to do with the module state. Perhaps we should rename the "SRE state" to `sre_state` (or `re_state`) in order to make the distinction more explicit.

erlend-aasland replied:
> * name the state after the module: `_sremodulestate`

Fixed in 1c925b9

erlend-aasland replied:
> * name the state variable in functions `state`, not `module_state`

@tiran, what do you think of the proposal above? (Renaming existing SRE state variables to `sre_state` or `re_state`, in order to make the distinction explicit.) It will explode the number of lines in this PR, but if you're fine with that, I'll get to it.

tiran reviewed (changes requested):
See inline comments

While you are working on the module, please remove ``PyModule_GetDict()`` and replace the code with ``PyModule_AddStringConstant()``.

Commit:
Address Heimes' review: improve module state naming

Commit:
Merge GH-23101

Commit:
Update clinic

Commit:
Address Heimes' review: Drop using _PyType_GetModuleByDef

Commit:
Adjust idlelib tiptest

from Lib/idlelib/idle_test/test_calltip.py:
The tests of builtins may break if inspect or the docstrings change,
but a red buildbot is better than a user crash (as has happened).
For a simple mismatch, change the expected output to the actual.

tiran reviewed (commented):
Thanks!

I'm leaving the final ack to @corona10 and @vstinner

corona10 reviewed (dismissed):
lgtm awesome,

I ran manually some tests and works well.

corona10 reviewed (commented):
But I want to know why this behavior is changed.

## master
```python
>>> import re
>>> p = re.compile('')
>>> p.sub.__doc__
'Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.'
```

## PR
```python
>>> import re
>>> p = re.compile('')
>>> p.sub.__doc__
None
```

Commit:
Revert "Adjust idlelib tiptest"

This reverts commit a6bc4810d6eb4551d192db90b6779f96a1cf47ab.

terryjreedy commented on Lib/idlelib/idle_test/test_calltip.py:
```diff
@@ -88,10 +88,7 @@ class SB:  __call__ = None
-(repl, string, count=0)
-Return the string obtained by replacing the leftmost \
-non-overlapping occurrences o...''')
+        tiptest(p.sub, '(repl, string, count=0)')
```
tiptest calls idlelib.calltip.get_argspec(p.sub).  The latter creates a tip from p.sub's signature and __doc__ attribute.  I verified that with this patch, p.sub.__doc__ is None and that help(p.sub) ends with the sentence "Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.".  This sentence is p.sub.__doc__ in installed 3.10.0a2.

Help (pydoc) gets the docstring from `re.Pattern.sub.__doc__`.  IDLE no longer does, because `p.sub.__doc__` no longer gets looked-up on class function where it resides.  For a user class, `instance.method.__doc__` is `class.method.__doc__`.

This patch changes `type(p.sub)` from `builtin function or method` to `builtin function`.  Does that have anything to do with the behavior change?  It seems like a bug.

erlend-aasland replied:
> Help (pydoc) gets the docstring from `re.Pattern.sub.__doc__`. IDLE no longer does, because `p.sub.__doc__` no longer gets looked-up on class function where it resides. For a user class, `instance.method.__doc__` is `class.method.__doc__`.

Aha; thanks
 
> This patch changes `type(p.sub)` from `builtin function or method` to `builtin function`. Does that have anything to do with the behavior change? It seems like a bug.

I believe it does. As I noted, the `dbm` module has similar issues. It definitely seems like a bug, and it seems to be tied to the `METH_METHOD` flag. I've tried looking in `Objects/descrobject.c` and `Objects/methodobject.c` without luck (I'm on thin ice in that directory).

erlend-aasland replied:
@terryjreedy `re.Pattern.sub` has changed from `builtin function or method` to `builtin method` (as it is now defined with the `METH_METHOD` flag). I guess the bug is in IDLE, then? Should I apply a6bc481 again, and file an issue for IDLE?

erlend-aasland replied:
FYI, `inspect.getdoc` gets it right:
```
>>> import inspect
>>> import re
>>> p = re.compile('')
>>> inspect.getdoc(p.sub)
'Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.'
>>> inspect.getdoc(re.Pattern.sub)
'Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl.'
```

erlend-aasland replied:
This fixes call tips in IDLE, but the REPL is still unable to fetch the docstring:
```
diff --git a/Lib/idlelib/calltip.py b/Lib/idlelib/calltip.py
index 549e224015..87a963dc1f 100644
--- a/Lib/idlelib/calltip.py
+++ b/Lib/idlelib/calltip.py
@@ -185,10 +185,7 @@ def get_argspec(ob):
              if len(argspec) > _MAX_COLS else [argspec] if argspec else [])
 
     # Augment lines from docstring, if any, and join to get argspec.
-    if isinstance(ob_call, types.MethodType):
-        doc = ob_call.__doc__
-    else:
-        doc = getattr(ob, "__doc__", "")
+    doc = inspect.getdoc(ob)
     if doc:
         for line in doc.split('\n', _MAX_LINES)[:_MAX_LINES]:
             line = line.strip()
```

terryjreedy replied:
Thank your for pointing out getdoc.  It turns out that special-casing MethodType for getting the docstring is not needed for any existing test.  Using inspect.getdoc is the right way to go, as it gets docstrings more often.  In fact, it breaks two other tests by getting docstrings that are currently missed.   I switched to getdoc and adjusted those tests in new issue 42415, #23416.  After I merge when tests pass, update your master and update this with `git merge upstream/master`.  Then tiptest(p.sub,...) should pass.

erlend-aasland replied:
> Thank your for pointing out getdoc. It turns out that special-casing MethodType for getting the docstring is not needed for any existing test. Using inspect.getdoc is the right way to go, as it gets docstrings more often. In fact, it breaks two other tests by getting docstrings that are currently missed. I switched to getdoc and adjusted those tests in new issue 42415, #23416. After I merge when tests pass, update your master and update this with `git merge upstream/master`. Then tiptest(p.sub,...) should pass.

Great, thanks!

Commit:
Merge branch 'master' into bpo-1635741/sre

corona10 reviewed (approved):
lgtm
